lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,232 @@
+        overlay.classList.add('lh-element-screenshot__overlay');
+        const elementRectInScreenshotCoords = {
+          width: Number(el.getAttribute('rectWidth')),
+          height: Number(el.getAttribute('rectHeight')),
+          left: Number(el.getAttribute('rectLeft')),
+          right: Number(el.getAttribute('rectRight')),
+          top: Number(el.getAttribute('rectTop')),
+          bottom: Number(el.getAttribute('rectBottom')),
+        };
+        overlay.appendChild(ElementScreenshotRenderer.render(
+          dom,
+          templateContext,
+          elementRectInScreenshotCoords,
+          fullPageScreenshot,
+          renderContainerSizeInDisplayCoords
+        ));
+        overlay.addEventListener('click', () => {
+          overlay.remove();
+        });
+
+        reportEl.appendChild(overlay);
+      });
+    }
+  }
+
+  /**
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords
+   * @param {Size} renderContainerSizeInDisplayCoords
+   * @return {number}
+   */
+  static _computeZoomFactor(elementRectInScreenshotCoords, renderContainerSizeInDisplayCoords) {
+    const targetClipToViewportRatio = 0.75;
+    const zoomRatioXY = {
+      x: renderContainerSizeInDisplayCoords.width / elementRectInScreenshotCoords.width,
+      y: renderContainerSizeInDisplayCoords.height / elementRectInScreenshotCoords.height,
+    };
+    const zoomFactor = targetClipToViewportRatio * Math.min(zoomRatioXY.x, zoomRatioXY.y);
+    return Math.min(1, zoomFactor);
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {ParentNode} templateContext
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords Region of screenshot to highlight.
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   * @param {Size} renderContainerSizeInDisplayCoords
+   * @return {Element}
+   */
+  static render(dom, templateContext, elementRectInScreenshotCoords,
+      fullPageScreenshot, renderContainerSizeInDisplayCoords) {
+    //
+    const tmpl = dom.cloneTemplate('#tmpl-lh-element-screenshot', templateContext);
+    const containerEl = dom.find('.lh-element-screenshot', tmpl);
+
+    containerEl.setAttribute('rectWidth', elementRectInScreenshotCoords.width.toString());
+    containerEl.setAttribute('rectHeight', elementRectInScreenshotCoords.height.toString());
+    containerEl.setAttribute('rectLeft', elementRectInScreenshotCoords.left.toString());
+    containerEl.setAttribute('rectRight', elementRectInScreenshotCoords.right.toString());
+    containerEl.setAttribute('rectTop', elementRectInScreenshotCoords.top.toString());
+    containerEl.setAttribute('rectBottom', elementRectInScreenshotCoords.bottom.toString());
+
+    // Zoom out when highlighted region takes up most of the viewport.
+    // This provides more context for where on the page this element is.
+    const zoomFactor =
+      this._computeZoomFactor(elementRectInScreenshotCoords, renderContainerSizeInDisplayCoords);
+
+    renderContainerSizeInDisplayCoords = {
```

Comment:
this is one of those derived pieces of state I was talking about and one of the reasons I didn't like mixing it in with `viewport` :) WDYT about...

```suggestion
    const elementPreviewSizeInScreenshotCoords = {
```

you can then create 

```js
const elementPreviewSizeInDisplayCoords = {
  width: elementPreviewSizeInScreenshotCoords * zoomFactor,
  height: elementPreviewSizeInScreenshotCoords * zoomFactor,
}
```

 and eliminate all of those inline `* zoomFactor` down below if you want